### PR TITLE
KVM: fix SSVM starting when overprovisioning memory

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR bypasses overprovisioning settings on memory for SSVMs to make sure these can always start.

In addition this PR adds autodeflate to memballoon on KVM VMs and sets the memory to `memory` instead of `currentMemory`. It implement sugestions by @rajujith at https://github.com/apache/cloudstack/issues/7430#issuecomment-1600245780. This solution seems not to fixs overprovisioning issues in VMs and a new PR will have to be created for a definite solution on that.

Fixes: #7430

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java [hunk 1]
```diff
@@ -282,7 +282,7 @@ public void setMemBalloning(boolean memoryBalloning) {
         @Override
         public String toString() {
             StringBuilder response = new StringBuilder();
-            response.append(String.format("<memory>%s</memory>\n", this.currentMemory));
+            response.append(String.format("<memory>%s</memory>\n", this.memory));
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java:285]: this seems to have no effect in my tests. looking for the answer.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by rajujith — APPROVED** [review]
LGTM. Fixes the VR /SystemVM memory over-provisioning issue. VR is not memory over-provisioned now. Default offering memory value is used as is by the VR irrespective of the memory.

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Review by rohityadavcloud — APPROVED** [review]
LGTM - though not sure about side effects. As far as I understand, this is the case for all systemvms/user-vms when memory is overcommitted. @DaanHoogland have you checked the behaviour when memory is overcommited for say VMware or XCP-ng/XS?

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by kiranchavala** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/hypervisor/KVMGuru.java [hunk 2]
```diff
@@ -213,7 +213,7 @@ protected void configureVmMemoryAndCpuCores(VirtualMachineTO virtualMachineTo, H
         Integer maxHostCpuCore = max.second();
 
         long minMemory = virtualMachineTo.getMinRam();
-        Long maxMemory = minMemory;
+        Long maxMemory = virtualMachineTo.getMaxRam();
```
Comment (rohityadavcloud) [inline server/src/main/java/com/cloud/hypervisor/KVMGuru.java:216]: I think only this fix was enough :)

Comment (rohityadavcloud) [reply server/src/main/java/com/cloud/hypervisor/KVMGuru.java:216]: Closing my PR as this main line/change has been addressed - https://github.com/apache/cloudstack/pull/7807

**Pull request merged**